File: docs/expressiveness/iceberg_refactor.ml
(* The "ultimate refactoring tool": add a new order kind, and the compiler points at EVERY
   match that no longer handles all cases — before any test runs. Reproduce:
     ocamlopt -c iceberg_refactor.ml
   (warning 8: this pattern-matching is not exhaustive; "Iceberg _" is unmatched). *)

type live = { id : int; price : int; remaining : int }
type done_reason = Filled | Cancelled
type order =
  | Live of live
  | Done of done_reason
  | Iceberg of { shown : int; hidden : int } (* NEW: only `shown` is visible to the book *)

(* This function predates `Iceberg` and was not updated — the compiler flags it for us. *)
let status_label = function
  | Live l -> Printf.sprintf "live(%d@%d)" l.remaining l.price
  | Done Filled -> "filled"
  | Done Cancelled -> "cancelled"

File: docs/expressiveness/iceberg_refactor.rs
// Same refactor in Rust: add a variant, and `match` fails to compile (error E0004) until every
// site handles it. Same safety guarantee as OCaml. Reproduce:
//   rustc --edition 2021 --crate-type lib iceberg_refactor.rs

pub enum Order {
    Live { id: u32, price: u32, remaining: u32 },
    Filled,
    Cancelled,
    Iceberg { shown: u32, hidden: u32 }, // NEW
}

pub fn status_label(o: &Order) -> String {
    match o {
        Order::Live { price, remaining, .. } => format!("live({remaining}@{price})"),
        Order::Filled => "filled".to_string(),
        Order::Cancelled => "cancelled".to_string(),
    }
}
